Create the Why Elixir blogpost (#29)
anoma/anoma-blog: mariari/why-elixir.md

> # Table of Contents

Reviewer: You can remove TOC. A TOC will autopopulate on the left side of the article. Before "Anoma: Why Elixir?" you need an excerpt. The excerpt will appear underneath the title and author information but above the media. See this post as an example. 

https://anoma.net/research/superspartan-by-hand

here is the syntax you want:

---
title: 
publish_date: 
category: 
image: 
excerpt: 

---

1.  [Anoma: Why Elixir?](#org9b02694)
    1.  [Live Interactive Systems](#org477e033)
    2.  [Designing Robust Systems](#org2c1931f)
    3.  [Specs: Or What is an Engine.](#orgb11b7eb)
    4.  [Conclusion](#org8571bee)

<a id="org9b02694"></a>

# Anoma: Why Elixir?

Anoma is a decentralized operating system with heterogeneous

> trust. For developing Anoma we've We've been using [Elixir](https://elixir-lang.org/). This choice

Reviewer: ```suggestion
trust. For developing Anoma we've been using [Elixir](https://elixir-lang.org/). This choice
```

is somewhat unconventional, and thus begs the question "why write
Anoma in Elixir!?". There are a plethora of reasons as to why [Elixir](https://elixir-lang.org/)

> is a good choice, and to motivate the reasoning we will go first from

Reviewer: ```suggestion
is a good choice. To motivate the reasoning, we will go first from
```

generic reasoning as to why it's a good fit for many projects before
diving deep into why it makes sense for Anoma in particular.

<a id="org477e033"></a>

## Live Interactive Systems

When building any large complicated system, the number one feature one
wants is to be able to understand how the system is operating along
with being able to determine if there are mishaps when carrying out a
design. For most software projects, this means sitting down and
thinking about the system abstractly or tracing through the system
execution in the case of direct questions. Thankfully for us, we have

> chosen [Elixir](https://elixir-lang.org/), which rest on the foundations of Erlang/OTP. In

Reviewer: ```suggestion
chosen [Elixir](https://elixir-lang.org/), which rests on the foundations of Erlang/OTP. In
```

practical terms, this means we get a live interactive system from
which we can ask the system questions.

For example, instead of having to sit and think about the
architecture of how the components of Anoma fit together, I can ask
the underlying system, "please give me the layout of processes for my
application":

![img](https://pomf2.lain.la/f/wqg712cm.png)

> Knowing the specifics of this diagram doesn't matter, however what

Reviewer: ```suggestion
Knowing the specifics of this diagram doesn't matter. However what
```

> does matter is that we can invite the system we are building into help

Reviewer: i think I know what you are trying to say - the system can help explain itself. The word choice here is a bit difficult to parse.

Author: Rewrote it to say this instead:

The specifics of the diagram doesn't matter. What does matter is that
the system can reflect on itself, helping to explain the very system
that we are building. Meaning, that the system we are building can be
in dialog with the builders to shape discussion around itself. No
longer is the system simply speculated about from the hazy
recollections that the engineers who made it have. A great upshot of
this is that if we have further questions about the system, we can
drill down and ask it more questions. For example what if we were
curious about one of those nodes in the diagram, what information can
we get from there?:

explaining itself. And if we have further questions about the system,
we can drill down and ask it more questions. For example what if we

> was curious about one of those nodes in the diagram, what information

Reviewer: ```suggestion
were curious about one of those nodes in the diagram, what information
```

> can Ι get from there?:

Reviewer: ```suggestion
can we get from there?:
```

        itself. Without inroads to the system, an interactive system
        would quickly turn opaque!
3.  The entire system available (from parser, to compiler, to
    visualization suite) from the user shell
    -   This allows a skilled engineer the ability to diagnose problems
        and see practical implications of what they are doing. This is
        apart of the [same technology that allowed people to debug a space
        probe 100 million miles away!](https://flownet.com/gat/jpl-lisp.html)
4.  A lack of artificial distinction between system and language.
    -   This means that the core functionality of the system is realized
        through the language itself, proving itself capable enough to
        compose a highly robust system.

With these properties and features laid out, this make Elixir quite
different from language in common use (Java, Rust, Python etc.)

With that said, there are [genres of languages](https://wiki.c2.com/?ImageBasedLanguage) that exhibit the same
features some of which have [superior visualization techniques](https://gtoolkit.com/). What is
it about the Erlang/OTP system that makes them special compared to

> them?

Reviewer: not sure what the second "them" is referring to - other languages?

Author: yeah those other languages

2.  The architecture of the application is split into a series of
    processes, each process are often referred to as actors.
    -   Actors are interesting objects as here are their central properties:
        1.  They contain some state they maintain
        2.  They can send and receive messages
        3.  Messages are processed sequentially (order is guaranteed)
            between the same two actors!
3.  Processes are isolated. Meaning that if a process crashes, it does
    not affect other processes. [Erlang the movie](https://www.youtube.com/watch?v=BXmOlCy0oBM) gives a good example
    of this at play on their real systems!
4.  Processes are managed by a chain of [supervisors](https://www.erlang.org/doc/apps/stdlib/supervisor.html).
    -   Supervisors are interesting in that they look after many
        processes (some of who may be supervisors). When a process they
        are supervising crashes, they have policies on what happens to
        the other processes under their control.
    -   It was realized that actors on their own are not enough to make a
        robust system, and this is a crucial component that is missing
        from most other actor models.

> 5.  A let it crash mentality is had

Reviewer: had?

Author: Rewrote the sentence to remove the had bit

    -   With Supervisors this is a very potent strategy for creating
        reliable and robust systems.

>         -   When crashes early happens, we can isolate the errors, this is

Reviewer: ```suggestion
        -   When crashes happen early, we can isolate the errors, this is
```

specification and those are the ones that will make users initially
interested and excited about Anoma. To understand what Anoma has to
offer, the user will look at the specification. If they are interested
in the presented promises of the system, they become incentivized to
take a proper look at the code and engage with it.

What the user is looking for is for an instantiation of the presented
proposal with desired properties. That is, they need to establish that
a piece of code lying inside a repository is an actual implementation
of the system described.

To make this easy, the engineers need to have a good interface that
makes it easy to establish a connection between concepts central to
the specification and our code. Otherwise, the application becomes
hard to adopt and develop for any outside party. Having such
properties can fundamentally depend on your choice of a language.

For example, if you are interested in functions, it matters whether
you present id : Bool -> Bool as lambda x -> x or {{true, true},
{false, false}}. There is a reason why we don't use set theory
everywhere in mathematics. If you think of functions as something that

> takes things in and pops things out, the type theoretic approach is

Reviewer 2: ```suggestion
take things in and pops things out, the type theoretic approach is
```

Author: takes is correct

probably easier to comprehend and work with.

> Folliwing this idea, a good requirement for the desired language is to

Reviewer 2: ```suggestion
Following this idea, a good requirement for the desired language is to
```

have a good first-class citizen corresponding to the central concept
of the specification: the [engine](https://specs.anoma.net/latest/node_architecture/engines.html?h=engine). Almost every functionality in the
Anoma infrastructure is specified as a functionality of a particular

> engine. Mempool functionality is done through the Mempool

Reviewer 2: ```suggestion
engine. Mempool behavior is implemented through the Mempool
```

> Engine. Transaction execition is done through the Execution Engine.

Reviewer 2: ```suggestion
Engine. Transaction execution is done through the Execution Engine.
```

---
title: Why Anoma

> publish_date:

Reviewer: - Can add a publish date - can be today / tomorrow. Won't update on the blog until the PR is merged. 
- For image you want to create a media folder then reference the item in image. I'm not sure what you have in mind. Typically, I use one of the AI image generating services to get image that gestures at the article. 
- For excerpt: Anoma is a decentralized operating system with heterogeneous
trust. For developing Anoma we've been using [Elixir](https://elixir-lang.org/). This choice is somewhat unconventional, and thus begs the question "why write Anoma in Elixir!?"
- For Category the choices are: compilers, distributed-systems, ecosystem, community, cryptography, applications, security, community
